File: solutions/two-pointers/018-4sum.py
I'll help you implement a solution for the 4Sum problem in Python. Here's a clean and efficient implementation with proper comments and error handling:

```python
#!/usr/bin/env python3
"""
4Sum Implementation - Find all unique quadruplets that sum to target

This module provides a solution to the 4Sum problem where we need to find
all unique quadruplets [nums[a], nums[b], nums[c], nums[d]] such that
nums[a] + nums[b] + nums[c] + nums[d] == target.

Time Complexity: O(n³) where n is the length of the input array
Space Complexity: O(1) excluding the space needed for output
"""

from typing import List


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        """
        Find all unique quadruplets in nums that sum up to target.
        
        Args:
            nums: List of integers
            target: Target sum
            
        Returns:
            List of lists containing unique quadruplets that sum to target
            
        Example:
            Input: nums = [1,0,-1,0,-2,2], target = 0
            Output: [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]
        """
        # Handle edge cases
        if not nums or len(nums) < 4:
            return []
        
        # Sort the array to handle duplicates and use two pointers approach
        nums.sort()
        n = len(nums)
        result = []
        
        # Fix first two numbers and use two pointers for the rest
        for i in range(n - 3):
            # Skip duplicates for first number
            if i > 0 and nums[i] == nums[i - 1]:
                continue
                
            # Early break if smallest possible sum is greater than target
            if nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target:
                break
                
            # Skip if largest possible sum is smaller than target
            if nums[i] + nums[n - 3] + nums[n - 2] + nums[n - 1] < target:
                continue
                
            for j in range(i + 1, n - 2):
                # Skip duplicates for second number
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                    
                # Use two pointers for the remaining two numbers
                left, right = j + 1, n - 1
                
                while left < right:
                    current_sum = nums[i] + nums[j] + nums[left] + nums[right]
                    
                    if current_sum == target:
                        result.append([nums[i], nums[j], nums[left], nums[right]])
                        
                        # Skip duplicates for third number
                        while left < right and nums[left] == nums[left + 1]:
                            left += 1
                        # Skip duplicates for fourth number
                        while left < right and nums[right] == nums[right - 1]:
                            right -= 1
                            
                        left += 1
                        right -= 1
                    elif current_sum < target:
                        left += 1
                    else:
                        right -= 1
        
        return result


def main():
    """
    Main function to test the 4Sum implementation
    """
    # Test cases
    solution = Solution()
    
    test_cases = [
        ([1, 0, -1, 0, -2, 2], 0),
        ([2, 2, 2, 2, 2], 8),
        ([-3, -2, -1, 0, 0, 1, 2], 0),
        ([], 0),  # Edge case: empty array
        ([1, 2, 3], 6),  # Edge case: array too small
    ]
    
    for nums, target in test_cases:
        print(f"\nInput: nums = {nums}, target = {target}")
        print(f"Output: {solution.fourSum(nums, target)}")


if __name__ == "__main__":
    main()
```

This implementation includes:

1. A complete solution to the 4Sum problem using a combination of sorting and two-pointer technique
2. Proper type hints and documentation
3. Efficient handling of duplicates to ensure unique quadruplets
4. Early breaking conditions to optimize performance
5. Edge case handling
6. A main function with test cases
7. Clear comments explaining the implementation
8. Proper Python conventions and best practices

The solution uses a two-pointer approach after fixing two numbers, which makes it more efficient than a brute force approach. The time complexity is O(n³) where n is the length of the input array.

The code includes optimizations such as:
- Early breaking when the smallest possible sum exceeds the target
- Skipping iterations when the largest possible sum is less than the target
- Handling duplicates at all levels to ensure unique quadruplets
- Proper sorting to enable the two-pointer approach

You can run this file directly to see the results of the test cases, or import the Solution class to use it in other parts of your project.

File: solutions/two-pointers/alternatives/018-4sum.js
/**
 * 018. 4Sum
 * Medium
 *
 * This problem demonstrates key concepts in Two Pointers.
 *
 * SOLUTION EXPLANATION:
 *
 * INTUITION:
 * Extend the 3Sum approach by adding one more outer loop. Fix two numbers and use
 * two pointers to find the remaining two numbers. Sorting helps avoid duplicates
 * and enables the two-pointer technique.
 *
 * APPROACH:
 * 1. **Sort Array**: Enables two-pointer technique and easy duplicate handling
 * 2. **Fix First Two Elements**: Use two nested loops to fix nums[i] and nums[j]
 * 3. **Two Pointers**: Use left and right pointers to find pairs summing to target - nums[i] - nums[j]
 * 4. **Skip Duplicates**: Avoid duplicate quadruplets by skipping repeated values
 *
 * WHY THIS WORKS:
 * By sorting and fixing two numbers, we reduce 4Sum to a 2Sum problem which can be
 * efficiently solved with two pointers. Careful duplicate skipping ensures unique results.
 *
 * TIME COMPLEXITY: O(n³)
 * - O(n log n) for sorting
 * - O(n³) for three nested loops with two pointers
 * - Overall: O(n³)
 * SPACE COMPLEXITY: O(1)
 * - Excluding output space, only using constant extra space
 *
 * EXAMPLE WALKTHROUGH:
 * ```
 * Input: nums = [1,0,-1,0,-2,2], target = 0
 * After sorting: [-2,-1,0,0,1,2]
 * i=0, nums[i]=-2:
 *   j=1, nums[j]=-1, need sum=3:
 *     left=2(0), right=5(2): sum=2<3, move left
 *     left=3(0), right=5(2): sum=2<3, move left
 *     ...
 *   j=2, nums[j]=0, need sum=2:
 *     left=3(0), right=5(2): sum=2✓ → quadruplet [-2,0,0,2]
 * Continue for all combinations...
 * Output: [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]
 * ```
 *
 * EDGE CASES:
 * - Array length < 4: return []
 * - All same numbers: check if 4*nums[0] == target
 * - Integer overflow: use careful arithmetic
 */

/**
 * Main solution for Problem 018: 4Sum
 *
 * @param {number[]} nums - Array of integers
 * @param {number} target - Target sum
 * @return {number[][]} - Array of unique quadruplets that sum to target
 *
 * Time Complexity: O(n³)
 * Space Complexity: O(1)
 */
function solve(nums, target) {
    // Edge case: need at least 4 numbers
    if (!nums || nums.length < 4) {
        return [];
    }

    const result = [];
    const n = nums.length;

    // Sort the array to enable two-pointer technique and skip duplicates
    nums.sort((a, b) => a - b);

    for (let i = 0; i < n - 3; i++) {
        // Skip duplicate values for the first number
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }

        for (let j = i + 1; j < n - 2; j++) {
            // Skip duplicate values for the second number
            if (j > i + 1 && nums[j] === nums[j - 1]) {
                continue;
            }

            // Two pointers approach for the remaining array
            let left = j + 1;
            let right = n - 1;

            while (left < right) {
                const sum = nums[i] + nums[j] + nums[left] + nums[right];

                if (sum === target) {
                    result.push([nums[i], nums[j], nums[left], nums[right]]);

                    // Skip duplicates for the third number
                    while (left < right && nums[left] === nums[left + 1]) {
                        left++;
                    }
                    // Skip duplicates for the fourth number
                    while (left < right && nums[right] === nums[right - 1]) {
                        right--;
                    }

                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
    }

    return result;
}

/**
 * Test cases for Problem 018: 4Sum
 */
function testSolution() {
    console.log('Testing 018. 4Sum');

    // Helper function to compare arrays of arrays
    const arraysEqual = (a, b) => {
        if (a.length !== b.length) return false;
        const sortedA = a.map(arr => arr.slice().sort((x, y) => x - y)).sort();
        const sortedB = b.map(arr => arr.slice().sort((x, y) => x - y)).sort();
        return JSON.stringify(sortedA) === JSON.stringify(sortedB);
    };

    // Test case 1: Example from problem
    const result1 = solve([1, 0, -1, 0, -2, 2], 0);
    const expected1 = [[-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]];
    console.assert(arraysEqual(result1, expected1), `Test 1 failed: expected ${JSON.stringify(expected1)}, got ${JSON.stringify(result1)}`);
    console.log(`Test 1 passed: fourSum([1,0,-1,0,-2,2], 0) = ${JSON.stringify(result1)}`);

    // Test case 2: No solution
    const result2 = solve([2, 2, 2, 2, 2], 8);
    const expected2 = [[2, 2, 2, 2]];
    console.assert(arraysEqual(result2, expected2), `Test 2 failed: expected ${JSON.stringify(expected2)}, got ${JSON.stringify(result2)}`);
    console.log(`Test 2 passed: fourSum([2,2,2,2,2], 8) = ${JSON.stringify(result2)}`);

    // Test case 3: Edge case - less than 4 elements
    const result3 = solve([1, 2, 3], 6);
    const expected3 = [];
    console.assert(arraysEqual(result3, expected3), `Test 3 failed: expected ${JSON.stringify(expected3)}, got ${JSON.stringify(result3)}`);
    console.log(`Test 3 passed: fourSum([1,2,3], 6) = ${JSON.stringify(result3)}`);

    // Test case 4: Another example
    const result4 = solve([2, 2, 2, 2, 2], 9);
    const expected4 = [];
    console.assert(arraysEqual(result4, expected4), `Test 4 failed: expected ${JSON.stringify(expected4)}, got ${JSON.stringify(result4)}`);
    console.log(`Test 4 passed: fourSum([2,2,2,2,2], 9) = ${JSON.stringify(result4)}`);

    // Test case 5: Negative target
    const result5 = solve([-3, -1, 0, 2, 4, 5], 2);
    const expected5 = [[-3, -1, 2, 4]];
    console.assert(arraysEqual(result5, expected5), `Test 5 failed: expected ${JSON.stringify(expected5)}, got ${JSON.stringify(result5)}`);
    console.log(`Test 5 passed: fourSum([-3,-1,0,2,4,5], 2) = ${JSON.stringify(result5)}`);

    console.log('All test cases passed for 018. 4Sum!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 018. 4Sum ===');
    console.log('Category: Two Pointers');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on two pointers concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */
